temporal-sdk/src/test/java/io/temporal/internal/replay/OutdatedDirectQueryReplayWorkflowRunTaskHandlerTest.java

```diff
@@ -0,0 +1,194 @@
+import io.temporal.internal.common.WorkflowExecutionUtils;
+import io.temporal.internal.history.LocalActivityMarkerUtils;
+import io.temporal.internal.statemachines.ExecuteLocalActivityParameters;
+import io.temporal.internal.statemachines.WorkflowStateMachines;
+import io.temporal.internal.worker.SingleWorkerOptions;
+import io.temporal.testing.internal.SDKTestWorkflowRule;
+import io.temporal.worker.WorkflowImplementationOptions;
+import io.temporal.workflow.Workflow;
+import io.temporal.workflow.shared.TestActivities;
+import io.temporal.workflow.shared.TestWorkflows;
+import java.time.Duration;
+import java.util.function.BiFunction;
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+@Issue("https://github.com/temporalio/sdk-java/issues/1371")
+public class OutdatedDirectQueryReplayWorkflowRunTaskHandlerTest {
+  private static final String QUERY_RESULT = "queryIsDone";
+
+  @Rule
+  public SDKTestWorkflowRule testWorkflowRule =
+      SDKTestWorkflowRule.newBuilder()
+          .setWorkflowTypes(WorkflowWithLocalActivityInASecondWorkflowTask.class)
+          .setActivityImplementations(new TestActivities.TestActivitiesImpl())
+          .build();
+
+  private WorkflowStateMachines stateMachines;
+
+  @Test
+  public void queryIsOutdated() {
+    TestWorkflows.TestWorkflowWithQuery noArgsWorkflow =
+        testWorkflowRule.newWorkflowStub(TestWorkflows.TestWorkflowWithQuery.class);
+    noArgsWorkflow.execute();
+    // check that a normal query works
+    assertEquals("A normal query should work", QUERY_RESULT, noArgsWorkflow.query());
+    WorkflowExecution workflowExecution = WorkflowStub.fromTyped(noArgsWorkflow).getExecution();
+
+    WorkflowExecutionHistory workflowExecutionHistory =
+        testWorkflowRule.getExecutionHistory(workflowExecution);
+
+    PollWorkflowTaskQueueResponseOrBuilder wft =
+        PollWorkflowTaskQueueResponse.newBuilder()
+            .setWorkflowExecution(workflowExecution)
+            .setPreviousStartedEventId(
+                WorkflowExecutionUtils.getEventsOfType(
+                        workflowExecutionHistory.getHistory(),
+                        EventType.EVENT_TYPE_WORKFLOW_TASK_STARTED)
+                    .get(0)
+                    .getEventId())
+            .setHistory(workflowExecutionHistory.getHistory())
+            .setQuery(WorkflowQuery.newBuilder().setQueryType("some-query").build());
+
+    ReplayWorkflowRunTaskHandler handler =
+        new ReplayWorkflowRunTaskHandler(
+            "UnitTest",
+            createReplayWorkflow(workflowExecutionHistory),
+            wft,
+            SingleWorkerOptions.newBuilder().build(),
+            new NoopScope(),
+            mock(BiFunction.class));
+
+    stateMachines = handler.getWorkflowStateMachines();
+    QueryResult queryResult =
+        handler.handleDirectQueryWorkflowTask(
+            wft, new FullHistoryIterator(workflowExecutionHistory.getEvents()));
```

Comment:
Some comments about what would have gone wrong were this not fixed would be helpful

Reply:
The test has a reference to a github issue, the github issue has a write-up of what's happening. I stopped copying the long narratives into tests code as I hope we have github as a reliable source of truth (Somebody migrating on SVN in the future is not thanking me)

Reply:
Sure, that works